File: plover_vm/memory/ram.py
"""64 KiB logical RAM (2×32K A15 bank unified)."""

from __future__ import annotations

RAM_SIZE = 64 * 1024


class Ram64K:
    def __init__(self) -> None:
        self._mem = bytearray(RAM_SIZE)

    def read(self, addr: int) -> int:
        return self._mem[addr & 0xFFFF]

    def write(self, addr: int, val: int) -> None:
        self._mem[addr & 0xFFFF] = val & 0xFF

    def load(self, data: bytes | bytearray, base: int = 0) -> None:
        for i, b in enumerate(data):
            self.write(base + i, b)

    def snapshot(self) -> bytes:
        return bytes(self._mem)

File: crates/plover_mmu/src/ram.rs
pub const RAM_SIZE: usize = 64 * 1024;

#[derive(Clone, Debug, Default)]
pub struct Ram64K {
    mem: Vec<u8>,
}

impl Ram64K {
    pub fn new() -> Self {
        Self {
            mem: vec![0; RAM_SIZE],
        }
    }

    pub fn read(&self, addr: u16) -> u8 {
        self.mem[addr as usize]
    }

    pub fn write(&mut self, addr: u16, val: u8) {
        self.mem[addr as usize] = val;
    }

    pub fn load(&mut self, data: &[u8], base: u16) {
        for (i, &b) in data.iter().enumerate() {
            let addr = base.wrapping_add(i as u16);
            self.write(addr, b);
        }
    }

    pub fn nonzero_count(&self) -> usize {
        self.mem.iter().filter(|&&b| b != 0).count()
    }
}
